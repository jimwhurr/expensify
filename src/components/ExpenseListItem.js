import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment'
import numeral from 'numeral';
import 'numeral/locales/en-gb';

numeral.locale('en-gb');

const ExpenseListItem = ( { id, description, amount, createdAt }) => (
    <div>
        <h3><Link to={`/edit/${id}`}>{description}</Link></h3>
        <p>
        {numeral(amount/100).format('$0,0.00')}
        -
        {moment(createdAt).format('Do MMMM YYYY')}
        </p>
    </div>
);

export default ExpenseListItem;