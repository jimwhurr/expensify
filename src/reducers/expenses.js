// -------------------------------------------------------------------------
//                     E X P E N S E S   R E D U C E R
// vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv

const expensesReducerDefaultState = [];
export default (state = expensesReducerDefaultState, action) => {
    switch(action.type) {

        case 'ADD_EXPENSE':     // REMEMBER we DON'T want to change the prev state!
            return [
                ...state,
                action.expense
            ];

        case 'REMOVE_EXPENSE':
            return state.filter( ({ id }) => id !== action.id );

        case 'EDIT_EXPENSE':
            return state.map( (expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates   // spread overrides
                    }
                }
                else {
                    return expense;
                }
            });

        case 'SET_EXPENSES':
            return action.expenses;
            
        default:
            return state;
    }
};
