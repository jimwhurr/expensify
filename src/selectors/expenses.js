import moment from 'moment';

// -------------------------------------------------------------------------
//                G E T   V I S I B L E   E X P E N S E S
// vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv

export default (expenses, { text, sortBy, startDate, endDate }) => {
    // console.log('params: {', text, sortBy, startDate, endDate, '}');
    return expenses.filter((expense) => {

        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ?  startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day'): true;
        
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;

    }).sort( (a, b) => {
        if (sortBy === 'date') {
            return (a.createdAt < b.createdAt) ? 1 : -1;
        }
        else if (sortBy === 'amount') {
            // console.log('a: ', a, ' b: ', b, a < b);
            return (a.amount < b.amount) ? 1 : -1;
        }
    });
};