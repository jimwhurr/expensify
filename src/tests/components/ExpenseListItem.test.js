import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

// <ExpenseListItem key={expense.id} {...expense} />            

describe('ExpenseListItemu', () => {

    test('should render expense list with given expenses', () => {
        const expense = expenses[0];
        const wrapper = shallow(<ExpenseListItem {...expense} />);
        expect(wrapper).toMatchSnapshot();
    });
});