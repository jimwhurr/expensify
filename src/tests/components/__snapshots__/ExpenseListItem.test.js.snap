// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseListItemu should render expense list with given expenses 1`] = `
<div>
  <h3>
    <Link
      replace={false}
      to="/edit/1"
    >
      Gum
    </Link>
  </h3>
  <p>
    Â£195.00
    -
    1st January 1970
  </p>
</div>
`;
